vm/devices/net/netvsp/src/lib.rs

```diff
@@ -2350,6 +2332,51 @@ impl<T: RingMem> NetChannel<T> {
         Ok(is_packet)
     }
 
+    /// Process RNDIS packet messages, which may contain multiple RNDIS packets
+    /// in a single vmbus message.
+    ///
+    /// On entry, the reader has already read the RNDIS message header of the
+    /// first RNDIS packet in the message.
+    fn handle_rndis_packet_messages(
+        &mut self,
+        buffers: &ChannelBuffers,
+        state: &mut ActiveState,
+        id: TxId,
+        mut message_len: usize,
+        mut reader: PacketReader<'_>,
+        segments: &mut Vec<TxSegment>,
+    ) -> Result<usize, WorkerError> {
+        // There may be multiple RNDIS packets in a single message, concatenated
+        // with each other. Consume them until there is no more data in the
+        // RNDIS message.
+        let mut num_packets = 0;
+        loop {
+            let next_message_offset = message_len
+                .checked_sub(size_of::<rndisprot::MessageHeader>())
+                .ok_or(WorkerError::RndisMessageTooSmall)?;
+
+            self.handle_rndis_packet_message(
+                id,
+                reader.clone(),
+                &buffers.mem,
+                segments,
+                &mut state.stats,
+            )?;
+            num_packets += 1;
+
+            reader.skip(next_message_offset)?;
+            if reader.len() == 0 {
+                break;
+            }
+            let header: rndisprot::MessageHeader = reader.read_plain()?;
+            if header.message_type != rndisprot::MESSAGE_TYPE_PACKET_MSG {
```

Comment:
just for my understanding, is it a requirement that only a packet message can follow a previous packet message? I didnt see anything in https://learn.microsoft.com/en-us/windows-hardware/drivers/network/remote-ndis-packet-msg

Reply:
That's the way vmswitch has always behaved, yeah. It's not an rndis protocol thing, it's a vmswitch behavior.

vmswitch actually limits you to 8 packets per message. I didn't bother with that restriction.